xds/internal/balancer/cdsbalancer/cdsbalancer_test.go

```diff
@@ -315,507 +322,1272 @@ func setupWithWatch(t *testing.T) (*fakeclient.Client, *cdsBalancer, *testEDSBal
+	if err != nil {
+		t.Fatalf("Failed to dial local test server: %v", err)
 	}
-	// Make sure the CDS balancer reports an error picker.
-	ctx, ctxCancel := context.WithTimeout(context.Background(), defaultTestTimeout)
-	defer ctxCancel()
+	t.Cleanup(func() { cc.Close() })
+
+	// Verify that the specified cluster resource is requested.
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
 	select {
 	case <-ctx.Done():
-		t.Fatalf("timeout when waiting for an error picker")
-	case picker := <-tcc.NewPickerCh:
-		if _, perr := picker.Pick(balancer.PickInfo{}); perr == nil {
-			t.Fatalf("CDS balancer returned a picker which is not an error picker")
+		t.Fatal("Timeout when waiting for cluster resource to be requested")
+	case <-cdsResourceRequestedCh:
+	}
+
+	// Push a resolver error that is not a resource-not-found error.
+	resolverErr := errors.New("resolver-error-not-a-resource-not-found-error")
+	r.ReportError(resolverErr)
+
+	// Ensure that the ClientConn moves to TransientFailure.
+	for state := cc.GetState(); state != connectivity.TransientFailure; state = cc.GetState() {
+		if !cc.WaitForStateChange(ctx, state) {
+			t.Fatalf("Timed out waiting for state change. got %v; want %v", state, connectivity.TransientFailure)
 		}
 	}
 
-	// Here we invoke the watch callback registered on the fake xdsClient. This
-	// will trigger the watch handler on the CDS balancer, which will attempt to
-	// create a new EDS balancer. The fake EDS balancer created above will be
-	// returned to the CDS balancer, because we have overridden the
-	// newChildBalancer function as part of test setup.
-	cdsUpdate := xdsresource.ClusterUpdate{
-		ClusterName: serviceName,
-		LBPolicy:    wrrLocalityLBConfigJSON,
+	// Drain the resolver error channel.
+	select {
+	case <-resolverErrCh:
+	default:
+	}
+
+	// Ensure that the resolver error is propagated to the RPC caller.
+	client := testgrpc.NewTestServiceClient(cc)
+	_, err = client.EmptyCall(ctx, &testpb.Empty{})
+	if code := status.Code(err); code != codes.Unavailable {
+		t.Fatalf("EmptyCall() failed with code: %v, want %v", code, codes.Unavailable)
+	}
+	if err != nil && !strings.Contains(err.Error(), resolverErr.Error()) {
+		t.Fatalf("EmptyCall() failed with err: %v, want %v", err, resolverErr)
 	}
-	wantCCS := edsCCS(serviceName, nil, false, wrrLocalityLBConfigJSON, noopODLBCfgJSON)
-	if err := invokeWatchCbAndWait(ctx, xdsC, cdsWatchInfo{cdsUpdate, nil}, wantCCS, edsB); err != nil {
+
+	// Also verify that the watch for the cluster resource is not cancelled.
+	sCtx, sCancel := context.WithTimeout(ctx, defaultTestShortTimeout)
+	defer sCancel()
+	select {
+	case <-sCtx.Done():
+	case <-cdsResourceCanceledCh:
+		t.Fatal("Watch for cluster resource is cancelled when not expected to")
+	}
+
+	// Start a test service backend.
+	server := stubserver.StartTestService(t, nil)
+	t.Cleanup(server.Stop)
+
+	// Configure cluster and endpoints resources in the management server.
```

Comment:
"Configure good cluster and endpoints" please

Reply:
Done.